File: backend/app/routes/snaps.py
"""Snap feed routes — thin HTTP layer."""

from fastapi import APIRouter, Depends, Query

from app.lib.auth_dependency import get_current_user
from app.services.snap import (
    complete_snap,
    list_snaps,
    save_snap,
)

router = APIRouter(prefix="/snaps", tags=["snaps"])


@router.get("/{business_id}")
async def list_snaps_route(
    business_id: str,
    tab: str = Query("attention"),
    limit: int | None = Query(None, ge=1),
    user: dict = Depends(get_current_user),
):
    """Return live Snaps for a tab: attention, recommendation, or priority."""
    snaps = await list_snaps(
        business_id=business_id,
        user_id=user["user_id"],
        tab=tab,
        limit=limit,
    )
    return {"snaps": snaps, "tab": tab, "count": len(snaps)}


@router.post("/{business_id}/{snap_id}/save")
async def save_snap_route(
    business_id: str,
    snap_id: str,
    user: dict = Depends(get_current_user),
):
    """Save a Snap, moving it to the priority tab."""
    snap = await save_snap(
        business_id=business_id,
        snap_id=snap_id,
        user_id=user["user_id"],
    )
    return {"snap": snap}


@router.post("/{business_id}/{snap_id}/complete")
async def complete_snap_route(
    business_id: str,
    snap_id: str,
    user: dict = Depends(get_current_user),
):
    """Mark a Snap completed, removing it from all tabs."""
    snap = await complete_snap(
        business_id=business_id,
        snap_id=snap_id,
        user_id=user["user_id"],
    )
    return {"snap": snap}

File: frontend/src/lib/services/snaps.ts
/**
 * Snap feed API service.
 *
 * The attention and recommendation tabs are served live from Redis. The
 * priority tab holds Snaps the user saved and is served from the database.
 */

import { request } from "./http";

export type SnapTab = "attention" | "recommendation" | "priority";

export type SnapType =
  | "recommendation"
  | "attention"
  | "warning"
  | "opportunity";

export type SnapPriority = "low" | "medium" | "high" | "critical";

export type SnapStatus = "active" | "pending" | "completed";

export type Snap = {
  snap_id: string;
  business_id: string;
  type: SnapType;
  priority: SnapPriority;
  confidence: number;
  title: string;
  message: string;
  why_it_matters: string;
  action: string;
  domain: string;
  status: SnapStatus;
  created_at: string | null;
};

type SnapListResponse = {
  snaps: Snap[];
  tab: SnapTab;
  count: number;
};

type SnapResponse = {
  snap: Snap;
};

/**
 * Fetch Snaps for a tab.
 * @param businessId - The business to fetch Snaps for
 * @param tab - Which tab to populate
 * @param limit - Maximum Snaps to return
 */
export async function listSnaps(
  businessId: string,
  tab: SnapTab,
  limit?: number,
): Promise<Snap[]> {
  const params = new URLSearchParams({ tab });
  if (limit) params.set("limit", String(limit));

  const { snaps } = await request<SnapListResponse>(
    `/snaps/${businessId}?${params}`,
    { silent: true },
  );
  return snaps;
}

/**
 * Save a Snap, moving it to the priority tab.
 */
export async function saveSnap(
  businessId: string,
  snapId: string,
): Promise<Snap> {
  const { snap } = await request<SnapResponse>(
    `/snaps/${businessId}/${snapId}/save`,
    { method: "POST" },
  );
  return snap;
}

/**
 * Mark a Snap completed, removing it from every tab.
 */
export async function completeSnap(
  businessId: string,
  snapId: string,
): Promise<Snap> {
  const { snap } = await request<SnapResponse>(
    `/snaps/${businessId}/${snapId}/complete`,
    { method: "POST" },
  );
  return snap;
}
